### PR TITLE
feat: add V4 position planner

### PR DESCRIPTION
**Pull request by malik672**
implement position planner

## Summary by CodeRabbit

- **New Features**
	- Introduced the `v4_position_planner` module, enhancing utilities for position planning in version 4 of the application.
	- Added the `V4PositionPlanner` struct, enabling users to manage liquidity and position actions, including minting, increasing, decreasing, and burning positions, as well as settling and taking pairs of tokens.
	- Added the `currency_map` module with a new function to convert currencies to their respective addresses.
	- Implemented the `to_address` function to facilitate address retrieval for native and wrapped currencies.
	- Updated dependencies to include `derive_more` for enhanced functionality.

### CODE REVIEW

File: src/utils/v4_position_planner.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        SettlePairParams, SweepParams, TakePairParams,
+    },
+    entities::Pool,
+};
+use alloy_primitives::{aliases::I24, Address, Bytes, U256};
+use uniswap_sdk_core::prelude::{BaseCurrency, Currency};
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct V4PositionPlanner {
+    /// Creates a new V4PositionPlanner instance
```
Comment (shuhuiluo) [inline src/utils/v4_position_planner.rs:14]:
```suggestion
```

File: src/utils/v4_position_planner.rs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            &pool.currency1,
+            pool.fee,
+            ticks,
+            pool.hooks,
+        )
+        .unwrap_or_default();
+
+        self.planner
+            .add_action(&Actions::MINT_POSITION(MintPositionParams {
+                poolKey: pool_key,
```
Comment (shuhuiluo) [inline src/utils/v4_position_planner.rs:45]:
```suggestion
        self.planner
            .add_action(&Actions::MINT_POSITION(MintPositionParams {
                poolKey: pool.pool_key.clone(),
```

File: src/utils/v4_position_planner.rs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+impl V4PositionPlanner {
+    #[allow(clippy::too_many_arguments)]
+    #[inline]
+    pub fn add_mint(
+        &mut self,
+        pool: &Pool,
+        tick_lower: i32,
+        tick_upper: i32,
```
Comment (shuhuiluo) [inline src/utils/v4_position_planner.rs:25]:
```suggestion
    pub fn add_mint<I: TickIndex>(
        &mut self,
        pool: &Pool,
        tick_lower: I,
        tick_upper: I,
```

File: src/utils/mod.rs [hunk 4]
```diff
@@ -4,6 +4,7 @@ pub mod price_tick_conversions;
 pub mod sorts_before;
 pub mod v4_base_actions_parser;
 pub mod v4_planner;
+pub mod v4_position_planner;
```
[the remark this answers is not in the document]
Comment (shuhuiluo) [reply src/utils/mod.rs:7]: Do it.

File: src/utils/v4_position_planner.rs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        }));
+    }
+
+    #[inline]
+    pub fn add_sweep(&mut self, currency: &Currency, recipient: Address) {
+        self.planner.add_action(&Actions::SWEEP(SweepParams {
+            currency: currency.address(),
+            recipient,
+        }));
+    }
```
Comment (shuhuiluo) [inline src/utils/v4_position_planner.rs:140]: Use `BaseCurrency` trait and [this](https://github.com/Uniswap/sdks/blob/main/sdks/v4-sdk/src/utils/currencyMap.ts).

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

File: src/utils/v4_position_planner.rs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    entities::Pool,
+};
+use alloy_primitives::{aliases::I24, Address, Bytes, U256};
+use uniswap_sdk_core::prelude::{BaseCurrency, Currency};
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct V4PositionPlanner {
+    /// Creates a new V4PositionPlanner instance
+    pub planner: V4Planner,
+}
```
Comment (shuhuiluo) [inline src/utils/v4_position_planner.rs:16]:
```suggestion
#[derive(Clone, Debug, Default, PartialEq, Deref)]
pub struct V4PositionPlanner(V4Planner);
```

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

**Review by shuhuiluo — CHANGES_REQUESTED** [review]
Fix the shit.

File: src/utils/v4_position_planner.rs [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        BurnPositionParams, DecreaseLiquidityParams, IncreaseLiquidityParams, MintPositionParams,
+        SettlePairParams, SweepParams, TakePairParams,
+    },
+    entities::Pool,
+};
+use alloy_primitives::{Address, Bytes, U256};
+use uniswap_sdk_core::prelude::BaseCurrency;
+use uniswap_v3_sdk::prelude::TickIndex;
+
+#[derive(Clone, Debug, Default, PartialEq)]
```
Comment (shuhuiluo) [inline src/utils/v4_position_planner.rs:13]:
```suggestion
#[derive(Clone, Debug, Default, PartialEq, Deref)]
```
Import `Deref` from `derive_more`.

File: src/utils/v4_position_planner.rs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+            currency0: currency0.address(),
+            currency1: currency1.address(),
+            recipient,
+        }));
+    }
+
+    #[inline]
+    pub fn add_sweep(&mut self, currency: &impl BaseCurrency, recipient: Address) {
+        self.0.add_action(&Actions::SWEEP(SweepParams {
+            currency: currency.address(),
```
Comment (shuhuiluo) [inline src/utils/v4_position_planner.rs:127]: [`currencyMap::toAddress`](https://github.com/Uniswap/sdks/blob/main/sdks/v4-sdk/src/utils/currencyMap.ts) is different from `BaseCurrency::address`. Implement it.

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

File: src/utils/currency_map.rs [hunk 9]
```diff
@@ -0,0 +1,10 @@
+use alloy_primitives::Address;
+use uniswap_sdk_core::prelude::BaseCurrency;
+
+#[inline]
+pub fn to_address(currency: &impl BaseCurrency) -> Address {
+    match currency.is_native() {
+        true => Address::ZERO,
+        false => currency.wrapped().address,
```
Comment (shuhuiluo) [inline src/utils/currency_map.rs:8]:
```suggestion
        false => currency.address(),
```

File: src/utils/v4_position_planner.rs [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+};
+use crate::{
+    abi::{
+        BurnPositionParams, DecreaseLiquidityParams, IncreaseLiquidityParams, MintPositionParams,
+        SettlePairParams, SweepParams, TakePairParams,
+    },
+    entities::Pool,
+};
+use alloy_primitives::{Address, Bytes, U256};
+use derive_more::derive::{Deref, DerefMut};
```
Comment (shuhuiluo) [inline src/utils/v4_position_planner.rs:13]:
```suggestion
use derive_more::{Deref, DerefMut};
```
Add `derive_more` in Cargo.toml.

File: src/utils/v4_position_planner.rs [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        }));
+    }
+
+    #[inline]
+    pub fn add_settle_pair(
+        &mut self,
+        currency0: &impl BaseCurrency,
+        currency1: &impl BaseCurrency,
+    ) {
+        self.0.add_action(&Actions::SETTLE_PAIR(SettlePairParams {
```
Comment (shuhuiluo) [inline src/utils/v4_position_planner.rs:104]:
```suggestion
        self.add_action(&Actions::SETTLE_PAIR(SettlePairParams {
```

File: src/utils/v4_position_planner.rs [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    #[inline]
+    pub fn add_take_pair(
+        &mut self,
+        currency0: &impl BaseCurrency,
+        currency1: &impl BaseCurrency,
+        recipient: Address,
+    ) {
+        self.0.add_action(&Actions::TAKE_PAIR(TakePairParams {
```
Comment (shuhuiluo) [inline src/utils/v4_position_planner.rs:117]:
```suggestion
        self.add_action(&Actions::TAKE_PAIR(TakePairParams {
```

**Review by shuhuiluo — CHANGES_REQUESTED** [review]

File: src/utils/v4_position_planner.rs (hunk 11 above)
Comment (malik672) [reply src/utils/v4_position_planner.rs:104]: this actually fails the linter

Comment (malik672) [reply src/utils/v4_position_planner.rs:104]: and it's valid

Comment (shuhuiluo) [reply src/utils/v4_position_planner.rs]: Because `derive_more` is not in Cargo.toml.

Comment (malik672) [reply src/utils/v4_position_planner.rs]: even when it's there, the linter still fails

**Pull request merged**